### PR TITLE
Improve unconstrained impl diagnostic (fixes #107295)

### LINKED ISSUE #107295: unconstrained type parameter makes no suggestions as to how to fix
### Code

```Rust
trait MyTrait {}

struct S;

impl <U> MyTrait for S {}
```

### Current output

```Shell
error[E0207]: the type parameter `U` is not constrained by the impl trait, self type, or predicates
 --> src/lib.rs:5:7
  |
5 | impl <U> MyTrait for S {}
  |       ^ unconstrained type parameter

For more information about this error, try `rustc --explain E0207`.
error: could not compile `playground` due to previous error
```

### Desired output

```Shell
error[E0207]: the type parameter `U` is not constrained by the impl trait, self type, or predicates
 --> src/lib.rs:5:7
  |
5 | impl <U> MyTrait for S {}
  |       ^ unconstrained type parameter

Either remove the type parameter (<U>) or make use of it, for example ` for S<U>`.

For more information about this error, try `rustc --explain E0207`.
error: could not compile `playground` due to previous error
```

### Rationale and extra context

If we have an unconstrained type parameter, we get an error stating that it is unconstrained, but no suggestions on what to do about it and force the user to google.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3ca555fca7f8220aaab497f87c26225b

Is there room to reduce the learning curve here with better diagnostics?

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by LlewVallis**
Fixes https://github.com/rust-lang/rust/issues/107295

### Problem

When a type/lifetime/const generic param is unconstrained ([as defined by the rules in this RFC](https://github.com/rust-lang/rfcs/blob/master/text/0447-no-unused-impl-parameters.md)) the current diagnostic isn't very helpful - it just tells you the param wasn't constrained. It isn't obvious (without having read the RFC) what can be done to *make* the param constrained.

### Solution

We add some additional notes to the existing diagnostic:
* In all cases, this PR adds an extra note which tells the user under what circumstances a generic is constrained. This message is slightly different depending on whether this is an impl trait or an inherent impl. The message slightly oversimplifies things to be more understandable, but the edge cases it misses are covered by the special messages below:
  * If the LHS was the thing being implemented (i.e. circular), then we use an extra message to explain that circularity isn't allowed
  * Otherwise, if the unconstrained param was mentioned in the RHS of an equality w/ a projection/associated-type then we add a message saying that this must mean the LHS of the equality has an unconstrained param

I added a few UI tests which probably explain these messages better than I have here :)

After looking at the RFC I linked, AFAICT there are no more edge cases where the dot points above could be incorrect, but I'm not an expert, so please correct me if there are.

<hr />

Also as a side note - the `enforce_impl_params_are_constrained` function is getting a bit long, lmk if you want me to split it up.

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? types

File: compiler/rustc_hir_analysis/src/impl_wf_check.rs [hunk 1]
```diff
@@ -183,6 +214,9 @@ fn report_unused_parameter(
     span: Span,
     kind: &str,
     name: Symbol,
+    is_impl_trait: bool,
+    is_bound_to_projection: bool,
+    is_bound_to_circular_projection: bool,
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/impl_wf_check.rs:219]: I haven't read the rest of your impl, this is just drive-by commentary.

Could please turn these `bool`s into custom enums? And merge these three flags as much as possible unless they are already mutually exclusive?

Comment (LlewVallis) [reply compiler/rustc_hir_analysis/src/impl_wf_check.rs]: Yep 👍 `is_impl_trait` is orthogonal to the other two booleans, but `is_bound_to_projection` and `is_bound_to_circular_projection` can be merged

File: compiler/rustc_hir_analysis/src/impl_wf_check.rs [hunk 2]
```diff
@@ -88,14 +88,36 @@ fn enforce_impl_params_are_constrained(
...
+        &mut constrained_parameters,
     );
 
+    // Reasons each generic is unconstrained, or Other if not present
+    let mut unconstrained_reasons = FxHashMap::default();
+    for (clause, _) in impl_predicates.predicates {
+        if let Some(projection) = clause.as_projection_clause() {
+            let mentioned_params = cgp::parameters_for(tcx, projection.term().skip_binder(), true);
+            let is_clause_circular =
+                Some(projection.required_poly_trait_ref(tcx).skip_binder()) == impl_trait_ref;
```
Comment (LlewVallis) [inline compiler/rustc_hir_analysis/src/impl_wf_check.rs:105]: Almost forgot, but wanted to add a quick note/question regarding `skip_binder` here... from what I understand, binders correspond to things like `for<'a>`, and `skip_binder` is *generally* a bad idea since it just unwraps the binder wo/ updating references to the quantifier.

My Q: is this specific usage safe? Or is some else more appropriate? (My understanding of this is a bit hazy 🙂)

**Comment by Dylan-DPC** [comment]
@LlewVallis if you can resolve the conflicts, we can push this forward for a review. Thanks

**Comment by LlewVallis** [comment]
Hey 👋 I'm short on time lately, so might leave this for someone else to pick up
